projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/Login2.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { Component } from 'react';
+import {requestAPI} from "../handler";
+import {Dialog, showDialog} from "@jupyterlab/apputils";
+import {
+    LOGIN2_BUTTON_CLASS,
+    LOGIN2_DIALOG_CLASS,
+    LOGIN2_INPUT_CLASS,
+    LOGIN_BUTTON_LABEL,
+} from "../utils";
+
+// Enum representing different states of authentication
+enum AuthStatus {
+    NONE,
+    WINDOW_OPENED,
+    SUCCESS,
+    FAILURE
+}
+
+interface ManualOAuth2FlowState {
+    authStatus: AuthStatus;
+}
+
+/**
+ * A React component to handle OAuth 2.0 authentication flow manually.
+ * This is not a common or ideal way to handle OAuth 2.0 authentication flow in real-world applications.
+ * In some authorization providers there's an issue with dynamic callback URLs.
+ * Since notebooks can be started using random URL we cannot speicfy fixed callback URL.
+ *
+ * This component provides an interactive method for users to authenticate
+ * through an external OAuth provider.
+ *
+ * Workflow:
+ * 1. An authentication window is opened for the user upon invoking `openAuthWindow`.
+ *    - This makes a request to the 'login' endpoint with the origin specified.
+ *    - The server responds with a redirect URL, which is then opened in a new window.
+ *
+ * 2. After the user completes the authentication on the OAuth provider's site,
+ *    they are redirected back, with an authorization code in the URL.
+ *
+ * 3. Users are instructed to copy and paste this URL into an input field manualy
+ *
+ * 4. This code is then sent to the 'login' endpoint to exchange it for tokens or finalize the authentication process.
+ *
+ * 5. The component's state is updated based on the success or failure of the authentication.
+ *
+ */
+export class ManualOAuth2Flow extends Component<{}, ManualOAuth2FlowState>  {
```

Comment:
Nit: Do we use class components throughout the jupyter plugin? I'm just curious if there's a reason behind it. Functional components with hooks read a lot better, i.e. you don't have to define a constructor, a render method and you don't have to define the initial state.

Reply:
I was wondering the same. 
We use class components through the jupyter plugin. So I am sticking to that in this change.

We are using React 15, it seems. Hooks and functional components were introduced in React 16 .
I have not delved into why yet. Duygu may shed more light on that. 

But we may change that for Beta , the very least jupyterlab 4 uses newer react.

Reply:
@antoniivanov @DeltaMichael we are using react 17, you can check the package.json. As for react 15 as far as i remember we were using a lower version of react when i started working for the integration for sure but not sure whether it was 15 or 16. At that point using components seemed ok, since most of the other extensions as git extension was using the same structure.
But as Antoni said we can think for changes for the beta version